Add Monaco Editor component and integrate with Nteract apps (#5267)
nteract/nteract: packages/monaco-editor/README.md

## Documentation

> ### Editor

Reviewer: Thanks for adding the detailed README for this!

Author: Excited for the OSS community to give it a try :)

> The `monaco-editor` provides the core functionality to render [Monaco Editor](https://microsoft.github.io/monaco-editor/) as a React component and also fetch completition items when connected to a Jupyter kernel in the context of a notebook. Since it is designed to work well with the semantics of a notebook, it requires the following props as laid down in the `IMonacoProps` interface:

Reviewer 2: ```suggestion
The `monaco-editor` package provides the core functionality to render [Monaco Editor](https://microsoft.github.io/monaco-editor/) as a React component. It also fetches code tab-completion items when running a notebook connected to a Jupyter kernel. To work well with the semantics of a notebook, the package requires the following props as laid down in the `IMonacoProps` interface:
```

> * `id` - A unique identifier for the editor instance

Author: Have updated the README.

> * `id` - A unique identifier for the editor instance. In the notebook context, since every cell is tied to a single instance of the editor, `id` refers to the unique ID of the cell.

Reviewer 2: This is a bit confusing. Is there one editor instance for an entire notebook or an editor instance for each cell?

Author: @willingc , just like code mirror, we'd have 1 editor instance for each cell.

* `contentRef` - A unique identifier for the editor's host application. In the notebook context, `contentRef` provides a reference to the container element for the main notebook app component.
* `theme` - Theme to be used for rendering the component ([docs](https://microsoft.github.io/monaco-editor/api/interfaces/monaco.editor.idiffeditorconstructionoptions.html#theme))

> * `language` - Valid language ID of a supported language (eg: `python`, `typescript`, `plaintext` etc.) Check out this Monaco Editor [playground](https://microsoft.github.io/monaco-editor/playground.html#extending-language-services-custom-languages) to add support for a language not yet supported out of the box

Reviewer 2: ```suggestion
* `language` - Valid language ID of a supported language (eg: `python`, `typescript`, `plaintext` etc.) Check out this Monaco Editor [playground](https://microsoft.github.io/monaco-editor/playground.html#extending-language-services-custom-languages) to add support for a language not yet supported out of the box.
```

> Besides the minimum required props to instantiate the component, we also provide support for a host of optional properties and handlers

Reviewer 2: ```suggestion
Besides the minimum required props to instantiate the component, we also provide support for a host of optional properties and handlers. These optional properties are found:
```

`options` - Specify a list of supported [EditorOptions](https://microsoft.github.io/monaco-editor/api/interfaces/monaco.editor.ieditoroptions.html) as key-value pairs when instantiating the component

Important callbacks:
* `onChange: (value: string, event?: any) => void` - Contents of the editor are changed.
* `onFocusChange: (focus: boolean) => void` - The Editor Component loses or gains focus
* `onCursorPositionChange: (selection: monaco.ISelection | null) => void` - Cursor position changes

### Completions

> The package also adds the capability to retrieve completion items from a connected Jupyter kernel. Completions are language specific token recommendations as the user attempts to type or enumerate the attributes of a class/object (usually with a `dot` operator). While we provide a default completion provider that works with the Jupyter kernel, we also support custom completion providers that users might want to register with their own language service.

Reviewer 2: ```suggestion
The package also adds the capability to retrieve code-completion items from a connected Jupyter kernel. Completions are language-specific token recommendations when the user attempts to type or enumerate the attributes of a class/object (usually with a `dot` operator and the `tab` completion key). While we provide a default completion provider that works with the Jupyter kernel, we also support custom completion providers that users might want to register with their own language service.
```

Reviewer 2: When you mention Jupyter kernel, is this any Jupyter kernel or just the Python kernel?

![jupyter completions](https://i.stack.imgur.com/rcieN.png)

The completion behavior is controlled by the following props:

* `enableCompletion` - Boolean flag to enable/disable the behavior entirely
* `shouldRegisterDefaultCompletion` - Boolean flag to enable/disable the default completion provider

> * `onRegisterCompletionProvider?: (languageId: string) => void` - Custom completion provider implementation for a Monaco Editor supported language.

Reviewer 2: For a Jupyter kernel like ipykernel does this need to be specified?

Author: We might not need to provide the `onRegisterCompletionProvider` but to enable completions for Ipykernel. We'd want to set `enableCompletion` to true and also set `shouldRegisterDefaultCompletion` to true

> Tip: If you want to enable completions in your app when you use this package, it is ideal to do it on a [`Web worker`](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers) which is separate from the UI thread. This provides a performance boost and ensures that the app doesn't stall UI updates when the editor is waiting for completions from the Jupyter Kernel. In Nteract, we use the [Monaco Editor Web pack plugin](https://github.com/microsoft/monaco-editor-webpack-plugin) to register and use the `editor` worker by Monaco. Check out the Monaco Editor [docs](https://github.com/microsoft/monaco-editor/blob/master/docs/integrate-esm.md) for more information on configuring the package and setting up web workers in a different set up.

Reviewer 2: ```suggestion
Performance Tip: If you want to enable completions in your app when you use this package, it is ideal to do it on a [`Web worker`](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers) which is separate from the UI thread. This provides a performance boost and ensures that the app doesn't stall UI updates when the editor is waiting for completions from the Jupyter Kernel. In Nteract, we use the [Monaco Editor Web pack plugin](https://github.com/microsoft/monaco-editor-webpack-plugin) to register and use the `editor` worker by Monaco. Check out the Monaco Editor [docs](https://github.com/microsoft/monaco-editor/blob/master/docs/integrate-esm.md) for more information on configuring the package and setting up web workers in a different set up.
```

> Additional perf improvements:

Reviewer 2: ```suggestion
Improving window resizing performance:
```
